File: app/src/main/java/com/atrainingtracker/trainingtracker/fragments/preferences/CloudUploadFragment.java
/*
 * aTrainingTracker (ANT+ BTLE)
 * Copyright (C) 2011 - 2019 Rainer Blind <rainer.blind@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see https://www.gnu.org/licenses/gpl-3.0
 */

package com.atrainingtracker.trainingtracker.fragments.preferences;

import android.content.SharedPreferences;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.preference.PreferenceManager;
import androidx.preference.PreferenceScreen;
import android.util.Log;

import com.atrainingtracker.BuildConfig;
import com.atrainingtracker.R;
import com.atrainingtracker.trainingtracker.TrainingApplication;
import com.dropbox.core.DbxRequestConfig;
import com.dropbox.core.android.Auth;
import com.dropbox.core.oauth.DbxCredential;

import java.util.LinkedList;
import java.util.List;

/**
 * Created by rainer on 01.02.16.
 */

public class CloudUploadFragment extends androidx.preference.PreferenceFragmentCompat
        implements SharedPreferences.OnSharedPreferenceChangeListener {
    private static final boolean DEBUG = TrainingApplication.getDebug(false);
    private static final String TAG = CloudUploadFragment.class.getName();

    @Nullable
    private PreferenceScreen mPSStrava, /* mPSRunkeeper, mPSTrainingPeaks, */ mPSEmailUpload;

    private SharedPreferences mSharedPreferences;

    private Boolean mAwaitDropboxResult = false;

    // private static DropboxAPI<AndroidAuthSession> mDBApi;

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        if (DEBUG) Log.i(TAG, "onCreatePreferences(savedInstanceState, rootKey=" + rootKey + ")");

        setPreferencesFromResource(R.xml.prefs, rootKey);

        mPSStrava = this.getPreferenceScreen().findPreference(TrainingApplication.PREFERENCE_SCREEN_STRAVA);
        // mPSRunkeeper = this.getPreferenceScreen().findPreference(TrainingApplication.PREFERENCE_SCREEN_RUNKEEPER);
        // mPSTrainingPeaks = this.getPreferenceScreen().findPreference(TrainingApplication.PREFERENCE_SCREEN_TRAINING_PEAKS);
        mPSEmailUpload = this.getPreferenceScreen().findPreference(TrainingApplication.PREFERENCE_SCREEN_EMAIL_UPLOAD);
    }

    @Override
    public void onResume() {
        super.onResume();
        if (DEBUG) Log.i(TAG, "onResume()");

        mPSStrava.setSummary(getPSStravaSummary());
        // mPSRunkeeper.setSummary(getPSRunkeeperSummary());
        // mPSTrainingPeaks.setSummary(getPSTrainingPeaksSummary());
        mPSEmailUpload.setSummary(getPSEmailUploadSummary());

        if (mAwaitDropboxResult) {
            DbxCredential dbxCredential = Auth.getDbxCredential();
            TrainingApplication.storeDropboxCredential(dbxCredential);
            mAwaitDropboxResult = false;
        }

        mSharedPreferences = PreferenceManager.getDefaultSharedPreferences(getActivity());
        mSharedPreferences.registerOnSharedPreferenceChangeListener(this);
    }

    @Override
    public void onPause() {
        super.onPause();

        // Unregister the listener whenever a key changes
        mSharedPreferences.unregisterOnSharedPreferenceChangeListener(this);
    }


    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        if (DEBUG) Log.i(TAG, "onSharedPreferenceChanged: key=" + key);

        if (TrainingApplication.SP_UPLOAD_TO_DROPBOX.equals(key)) {
            if (!TrainingApplication.uploadToDropbox()) { // -> Upload to Dropbox has been changed to false by the user
                TrainingApplication.deleteDropboxCredential();
            } else {                                      // -> Upload to Dropbox has been changed to true by the user
                Auth.startOAuth2PKCE(getActivity(), BuildConfig.DROPBOX_APP_KEY, new DbxRequestConfig(BuildConfig.DROPBOX_APP_KEY));
                mAwaitDropboxResult = true;
            }
        }
    }

    @NonNull
    private String getPSStravaSummary() {
        if (!TrainingApplication.uploadToStrava()) {
            return getString(R.string.prefsUploadToStravaSummary);
        } else {
            List<String> list = new LinkedList<>();
            if (TrainingApplication.uploadStravaGPS()) {
                list.add(getString(R.string.GPS));
            }
            if (TrainingApplication.uploadStravaAltitude()) {
                list.add(getString(R.string.altitude));
            }
            if (TrainingApplication.uploadStravaHR()) {
                list.add(getString(R.string.heart_rate));
            }
            if (TrainingApplication.uploadStravaPower()) {
                list.add(getString(R.string.power));
            }
            if (TrainingApplication.uploadStravaCadence()) {
                list.add(getString(R.string.cadence));
            }

            return getString(R.string.upload_to_strava_format, listToString(list, 5));
        }
    }

    @NonNull
    private String getPSRunkeeperSummary() {
        if (!TrainingApplication.uploadToRunKeeper()) {
            return getString(R.string.prefsUploadToRunkeeperSummary);
        } else {
            List<String> list = new LinkedList<>();
            if (TrainingApplication.uploadRunkeeperGPS()) {
                list.add(getString(R.string.GPS));
            }
            if (TrainingApplication.uploadRunkeeperHR()) {
                list.add(getString(R.string.heart_rate));
            }

            return getString(R.string.upload_to_runkeeper_format, listToString(list, 3));
        }
    }

    @NonNull
    private String getPSTrainingPeaksSummary() {
        if (!TrainingApplication.uploadToTrainingPeaks()) {
            return getString(R.string.prefsUploadToTrainingPeaksSummary);
        } else {
            List<String> list = new LinkedList<>();
            if (TrainingApplication.uploadTrainingPeaksGPS()) {
                list.add(getString(R.string.GPS));
            }
            if (TrainingApplication.uploadTrainingPeaksAltitude()) {
                list.add(getString(R.string.altitude));
            }
            if (TrainingApplication.uploadTrainingPeaksHR()) {
                list.add(getString(R.string.heart_rate));
            }
            if (TrainingApplication.uploadTrainingPeaksPower()) {
                list.add(getString(R.string.power));
            }
            if (TrainingApplication.uploadTrainingPeaksCadence()) {
                list.add(getString(R.string.cadence));
            }

            return getString(R.string.upload_to_training_peaks_format, listToString(list, 5));
        }
    }

    @NonNull
    private String getPSEmailUploadSummary() {
        if (!TrainingApplication.sendEmail()) {
            return getString(R.string.upload_via_email);
        } else {
            List<String> list = new LinkedList<>();
            if (TrainingApplication.sendTCXEmail()) {
                list.add(getString(R.string.TCX));
            }
            if (TrainingApplication.sendGPXEmail()) {
                list.add(getString(R.string.GPX));
            }
            if (TrainingApplication.sendCSVEmail()) {
                list.add(getString(R.string.CSV));
            }
            if (TrainingApplication.sendGCEmail()) {
                list.add(getString(R.string.GC));
            }

            if (list.isEmpty()) {
                return getString(R.string.send_email_only_summary_format, TrainingApplication.getSpEmailAddress());
            } else if (list.size() == 1) {
                return getString(R.string.send_email_format, list.get(0), TrainingApplication.getSpEmailAddress());
            } else {
                return getString(R.string.send_email_format_plural, listToString(list, 5), TrainingApplication.getSpEmailAddress());
            }
        }
    }

    private String listToString(@NonNull List<String> listOfString, int max) {
        int size = listOfString.size();
        if (size == max) {
            return getString(R.string.everything);
        } else if (size == 0) {
            return getString(R.string.summary);
        } else if (size == 1) {
            return listOfString.get(0);
        } else if (size == 2) {
            return getString(R.string.concatenate_2_format, listOfString.get(0), listOfString.get(1));
        } else {
            String lastOne = listOfString.get(size - 1);
            listOfString.remove(size - 1);

            StringBuilder result = new StringBuilder();
            for (String string : listOfString) {
                result.append(string);
                result.append(", ");
            }
            String firstPart = result.substring(0, result.length() - 2);

            return getString(R.string.concatenate_last_format, firstPart, lastOne);
        }
    }

}

File: app/src/main/java/com/atrainingtracker/trainingtracker/fragments/preferences/CloudUploadFragment.kt
/*
 * aTrainingTracker (ANT+ BTLE)
 * Copyright (c) 2011 - 2026 Rainer Blind <rainer.blind@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 */

package com.atrainingtracker.trainingtracker.fragments.preferences

import android.content.SharedPreferences
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.statusBarsPadding
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.ComposeView
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import androidx.preference.PreferenceFragmentCompat
import androidx.preference.PreferenceManager
import com.atrainingtracker.BuildConfig
import com.atrainingtracker.R
import com.atrainingtracker.trainingtracker.TrainingApplication
import com.atrainingtracker.trainingtracker.ui.theme.ATrainingTrackerTheme
import com.dropbox.core.DbxRequestConfig
import com.dropbox.core.android.Auth

class CloudUploadFragment : PreferenceFragmentCompat(), SharedPreferences.OnSharedPreferenceChangeListener {

    private var mSharedPreferences: SharedPreferences? = null
    private var mAwaitDropboxResult = false
    private var mHeaderComposeView: ComposeView? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mAwaitDropboxResult = savedInstanceState?.getBoolean(KEY_AWAIT_DROPBOX, false) ?: false
    }

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        if (DEBUG) Log.i(TAG, "onCreatePreferences(savedInstanceState, rootKey=$rootKey)")
        setPreferencesFromResource(R.xml.prefs_dropbox, null)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        val prefView = super.onCreateView(inflater, container, savedInstanceState)

        val root = LinearLayout(requireContext()).apply {
            orientation = LinearLayout.VERTICAL
            layoutParams = ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
        }

        mHeaderComposeView = ComposeView(requireContext()).apply {
            layoutParams = LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT)
            updateHeaderContent()
        }

        root.addView(mHeaderComposeView)
        prefView?.let { root.addView(it) }

        return root
    }

    private fun updateHeaderContent() {
        mHeaderComposeView?.setContent {
            ATrainingTrackerTheme {
                Column {
                    Surface(
                        color = MaterialTheme.colorScheme.primaryContainer
                    ) {
                        Column(
                            modifier = Modifier.statusBarsPadding()
                        ) {
                            Row(
                                modifier = Modifier
                                    .fillMaxWidth()
                                    .padding(horizontal = 16.dp, vertical = 4.dp),
                                verticalAlignment = Alignment.CenterVertically
                            ) {
                                Text(
                                    text = stringResource(R.string.Dropbox),
                                    style = MaterialTheme.typography.headlineSmall,
                                    color = MaterialTheme.colorScheme.onPrimaryContainer
                                )
                            }
                        }
                    }

                    // Connection Control (Buttons) - Matching Strava pattern
                    val isConnected = TrainingApplication.uploadToDropbox()
                    DropboxConnectionHeader(
                        isConnected = isConnected,
                        onConnectClick = {
                            Auth.startOAuth2PKCE(requireActivity(), BuildConfig.DROPBOX_APP_KEY, DbxRequestConfig(BuildConfig.DROPBOX_APP_KEY))
                            mAwaitDropboxResult = true
                        },
                        onDisconnectClick = {
                            TrainingApplication.deleteDropboxCredential()
                            TrainingApplication.setUploadToDropbox(false)
                            updateHeaderContent()
                        }
                    )
                }
            }
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        listView?.let { listView ->
            listView.clipToPadding = false
            androidx.core.view.ViewCompat.setOnApplyWindowInsetsListener(listView) { v, insets ->
                val systemBars = insets.getInsets(androidx.core.view.WindowInsetsCompat.Type.systemBars())
                v.setPadding(0, systemBars.top, 0, systemBars.bottom)
                insets
            }
        }
    }

    override fun onResume() {
        super.onResume()
        if (DEBUG) Log.i(TAG, "onResume()")

        if (mAwaitDropboxResult) {
            val dbxCredential = Auth.getDbxCredential()
            if (dbxCredential != null) {
                TrainingApplication.storeDropboxCredential(dbxCredential)
                TrainingApplication.setUploadToDropbox(true)
            }
            mAwaitDropboxResult = false
        }

        mSharedPreferences = PreferenceManager.getDefaultSharedPreferences(requireActivity())
        mSharedPreferences?.registerOnSharedPreferenceChangeListener(this)
        
        updateHeaderContent()
    }

    override fun onPause() {
        super.onPause()
        mSharedPreferences?.unregisterOnSharedPreferenceChangeListener(this)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putBoolean(KEY_AWAIT_DROPBOX, mAwaitDropboxResult)
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences?, key: String?) {
        if (DEBUG) Log.i(TAG, "onSharedPreferenceChanged: key=$key")

        if (TrainingApplication.SP_UPLOAD_TO_DROPBOX == key) {
            updateHeaderContent()
        }
    }

    companion object {
        private val TAG = CloudUploadFragment::class.java.name
        private val DEBUG = TrainingApplication.getDebug(false)
        private const val KEY_AWAIT_DROPBOX = "KEY_AWAIT_DROPBOX"
    }
}
